GetInfo() function

kaloyan-raev opened the pull request:
WIP of the first step for implementing Storj client in pure Go.

coyle commented on cli/cli.go:
```diff
@@ -0,0 +1,64 @@
+const version = "0.0.1"
+
+func main() {
+	// TODO support long flag names, i.e. --help, --version, etc.
```
You can use a CLI library to get some of this out of the box.

check out https://github.com/urfave/cli

coyle commented on cli/cli.go:
```diff
@@ -0,0 +1,64 @@
+	switch cmd {
+	case "get-info":
+		info, err := storj.GetInfo(env)
+		if err != nil {
```
idiomatic to do:
```Go
if err != nil {
    fmt.Fprintf(os.Stderr, "%v\n", err)
    return
} 

fmt.Printf("Title: %s\nDescription: %s\nVersion: %s\nHost: %s\n",info.Title, info.Description, info.Version, info.Host)
```

coyle commented on cli/cli.go:
```diff
@@ -0,0 +1,64 @@
+	"fmt"
+	"os"
+
+	"storj.io/storj"
```
where so you have these files in your path?

kaloyan-raev replied:
I made `$GOPATH/src/storj.io` to be a symlink to the github repo root. This way I simulate that the `storj.io/storj` is published and installed in a standard way.

coyle replied:
I would switch the import then to
```Go
storj github.com/Storj/go-libstorj
```

since once it gets merged that's what it will look like

kaloyan-raev replied:
I imagined there is a packaging system like in other languages and I can have a simple and stable package name like `storj.io/storj` and not one that depends on the hosting site and repo name. Isn't there any way to do this?

coyle replied:
There is a way to do vanity imports but we have to first set up some meta tags on our website. 
and if we did, then that would be
```Go
storj "storj.io/storj/go-libstorj"
```

coyle replied:
This is a good write up on how that works:
https://blog.bramp.net/post/2017/10/02/vanity-go-import-paths/

kaloyan-raev replied:
OK. I'll avoid vanity and stick to your initial suggestion:
```go
storj "github.com/Storj/go-libstorj"
```
So, this will be the way for clients to import the library. Is there are way for the cli to import the library in some "relative" way, so I don't need to create symlink to build successfully?

kaloyan-raev replied:
Found the way - I had to move the project's git repo to the correct location under GOPATH.

coyle replied:
The way I typically do this is I just work out of $GOPATH/github.com/<my-username>/go-libstorj on a separate branch then just fix the imports before creating a PR. Not sure if there is a better way.

coyle commented on storj/storj.go:
```diff
@@ -0,0 +1,78 @@
+
+// Env contains parameters for accessing the Storj network
+type Env struct {
+	URL string // TODO set DefaultURL as default value
```
This is typically done with an initialization function like
```Go
func NewEnv() Env {
    return Env{
         URL: DefaultURL,
    }
}
```

coyle commented on storj/storj.go:
```diff
@@ -0,0 +1,78 @@
+	infomap := m["info"]
+	v := infomap.(map[string]interface{})
+
+	info.Title = v["title"].(string)
```
If this doesn't contain title or title (for whatever odd reason) is not a string this will crash (panic).

For safety, this should first validate the existence of title. Something like:
```Go
info.title ok := v["title"].(string)
if !ok {
    return errors.New("invalid type for title")
}
```

coyle commented on storj/storj.go:
```diff
@@ -0,0 +1,78 @@
+	if err != nil {
+		return info, err
+	}
+	if resp.StatusCode != 200 {
```
It's idiomatic to use the http const instead of 200

```Go
if resp.StatusCode != http.StatusOK { 
```

coyle commented on storj/storj.go:
```diff
@@ -0,0 +1,78 @@
+		return info, fmt.Errorf("Unexpected response code: %d", resp.StatusCode)
+	}
+	b, err := ioutil.ReadAll(resp.Body)
+	resp.Body.Close()
```
move this up to after you get resp and do 
```Go
defer resp.Body.Close()
```

That way it is always closed no matter which return path you go.
For instance right now, if there is an open body and status code is not 200 there will be a leak.

coyle commented on storj/storj_test.go:
```diff
@@ -0,0 +1,84 @@
+func mockBridge() {
+	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
+		if r.URL.Path != "/" {
+			http.Error(w, fmt.Sprintf("Cannot GET %s", r.URL.Path), 404)
```
idiomatic Go to remove the else and return.

coyle commented on storj/storj_test.go:
```diff
@@ -0,0 +1,84 @@
+	time.Sleep(1000)
+}
+
+func TestGetInfo(t *testing.T) {
```
Check out 

https://dave.cheney.net/2013/06/09/writing-table-driven-tests-in-go

It's a really powerful pattern to write tests in Go

kaloyan-raev replied:
I re-implemented the tests following the table driven pattern. Please, have a look and let me know if I got it right?

kaloyan-raev commented:
@coyle Thanks for the detailed review! I believe I addressed all your comments. Let me know If I can make any other improvements.

One thing I am thinking about is splitting the storj.go file in two files:
- info.go - it will contain everything related to the get-info command
- common.go - it will contain common consts and structs like `DefaultURL` and `Env`

I'll do the same with storj_test.go. When I add new features, they will go into a separate file. This way I should have shorter files with clearer responsibilities. Everything will be still in one single package.

Is this a good practice in Go?
